docs/Hardware-Agnostic-Training.md

```diff
@@ -0,0 +1,67 @@
+
+```bash
+python src/levanter/main/train_lm.py \
+    --config_path config/gpt2_small.yaml \
+    --trainer.checkpointer.base_path output/gpt2_small_webtext \
+    --trainer.train_batch_size 256 \
+    --trainer.per_device_parallelism 128
+```
+This example command assumes you're starting your training run on 2 GPUs in a [FSDP (Fully Sharded Data Parallel)](https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html) setup. The `trainer.per_device_parallelism` parameter instructs Levanter to distribute the processing of each training batch across multiple accelerators, in this case, across 2 GPUs. With this setting, each GPU handles a subset of the training batch, processing 128 examples each, and their combined efforts contribute to the effective handling of the entire 256 batch size.
+
+We have comprehensive documentation on [getting started with training](Getting-Started-Training) in levanter, the [configuration file](Configuration-Guide.md), and [training on your own dataset](Training-On-Your-Data.md) that you should check out for more details. We also have a notebook tutorial on [how to add FSDP to custom architectures](https://colab.research.google.com/drive/1QX4yH3zRFF3Xiibf1aahETcSQ5nbcUMz) implemented in Levanter with Haliax.
+
+
+### How the Model Checkpoint is Saved
+The frequency of checkpoint saves can be adjusted by configuring the `checkpointer.save_interval` parameter in your training command. This parameter requires an integer value, which specifies the interval (in minutes) between each checkpoint save. By default, checkpoints are saved every 15 minutes.
+
+Whenever a checkpoint is saved, Levanter creates a new directory named after the training run ID within the path specified by `trainer.checkpointer.base_path`. This directory stores each checkpoint file. Every checkpoint file containes:
+
+* the model's parameters
+* the current optimizer and trainer states
+* meta data on the current training step and the time the training step was saved
+
+```
+user@DN0a2335e6 rj96a79n/step-1591 % ls
+metadata.json	model		training_state	worker-0.cert
+```
+
+### Moving the Checkpoint File to Google Storage
+To switch to a new hardware configuration mid training run, you need to move your checkpoint file and your training data over to the new hardware environemnt. If you started your training run on a GPU server and want to move it to a TPU VM instances, you can use the following command to upload your checkpoint file to a google storage drive that your TPU VM can access.
+
+```bash
+gsutil -m cp -r <GPU MODEL CHECKPT PATH> gs://<SOMEWHERE>
+```
+You need to be logged into your GCP account in your current environment for this command to run properly. Information on how to set this up can be found [here](Getting-Started-TPU-VM.md).
+
+### Re-Configuring `per_device_parallelism`
```

Comment:
again i think you can just leave it unspecified. you need to specify it if you need to do grad accumulation, but otherwise it should "just work"